### PR TITLE
[master]Set cache for volumes list of the host

### PR DESCRIPTION
**Pull request by trenxiaof**
Signed-off-by: Xiao Feng Ren <renxiaof@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by trenxiaof** [comment]
@bjhuangr @haoqf  please help review.

**Comment by trenxiaof** [comment]
@bjhuangr @SeanHQF please help review. have verified locally.

File: zvmsdk/hostops.py [hunk 1]
```diff
@@ -76,9 +77,12 @@ def guest_list(self):
             return guest_list
 
     def diskpool_get_volumes(self, pool_name):
-        diskpool_volume_list = self._smtclient.get_diskpool_volumes(pool_name)
-        with zvmutils.expect_invalid_resp_data(diskpool_volume_list):
-            return diskpool_volume_list
+        if self._volumes:
+            return self._volumes
```
Comment (jichenjc) [inline zvmsdk/hostops.py:81]: i don't have too much background of this PR
my question is if disk pool changes, what's the action to sync ?

before that question clear to me ,I want to hold this

Comment (SeanHQF) [reply zvmsdk/hostops.py:81]: I think disk pool change will cause the service restart and the cache cleaned

Comment (jichenjc) [reply zvmsdk/hostops.py:81]: no, adding disk is done in z/VM layer
how can we detect that?

Comment (trenxiaof) [reply zvmsdk/hostops.py:81]: > i don't have too much background of this PR
> my question is if disk pool changes, what's the action to sync ?
> 
> before that question clear to me ,I want to hold this

if the sdkserver not retart, the cache couldn't sync the latest disk_pool

Comment (SeanHQF) [reply zvmsdk/hostops.py:81]: Either request the user to restart zCC or add expire in cache

Comment (jichenjc) [reply zvmsdk/hostops.py:81]: >>>if the sdkserver not retart, the cache couldn't sync the latest disk_pool

that's the concern , if we have no other ways , we need well prepare for such case
add disk is something not rare..

Comment (SeanHQF) [reply zvmsdk/hostops.py:81]: Can we set this as a limitation since it only impacts onboarding?

File: zvmsdk/hostops.py [hunk 2]
```diff
@@ -37,6 +38,8 @@ class HOSTOps(object):
     def __init__(self):
         self._smtclient = smtclient.get_smtclient()
         self._volume_infos = {}
+        self._volumes = {}
```
Comment (SeanHQF) [inline zvmsdk/hostops.py:41]: should _volumes = None

**Comment by trenxiaof** [comment]
@jichenjc @SeanHQF  please help review. have tested locally

**Review by SeanHQF — APPROVED** [review]

**Comment by SeanHQF** [comment]
@trenxiaof pls solve the conflict, thanks

**Comment by SeanHQF** [comment]
@trenxiaof CI failed.

**Comment by SeanHQF** [comment]
@jichenjc can you help review? thanks

**Comment by jichenjc** [comment]
@bjhuangr can you help check this?

File: zvmsdk/hostops.py [hunk 3]
```diff
@@ -75,10 +78,26 @@ def guest_list(self):
...
+        if self._cache_enabled():
+            if (time.time() > self.cache_expiration):
+                self._volumes = None
+            if self._volumes:
+                return self._volumes
+            else:
+                self._volumes = self._smtclient.get_diskpool_volumes(pool_name)
+                self.cache_expiration = time.time() + \
+                    float(CONF.monitor.cache_interval)
+                with zvmutils.expect_invalid_resp_data(self._volumes):
```
Comment (bjhuangr) [inline zvmsdk/hostops.py:94]: with zvmutils.expect_invalid_resp_data() not helps

File: zvmsdk/hostops.py [hunk 4]
```diff
@@ -75,10 +78,26 @@ def guest_list(self):
...
+                    return self._volumes
+        else:
+            self._volumes = self._smtclient. \
+                get_diskpool_volumes(pool_name)
+            with zvmutils.expect_invalid_resp_data(self._volumes):
```
Comment (bjhuangr) [inline zvmsdk/hostops.py:99]: ditto

**Review by bjhuangr — CHANGES_REQUESTED** [review]

**Comment by bjhuangr** [comment]
@trenxiaof any plan to update this patch?

**Comment by trenxiaof** [comment]
> @trenxiaof any plan to update this patch?

Yes, will update it with your comments

**Comment by trenxiaof** [comment]
@huangrui add self.disk_pool to save the old disk_pool, if the disk_pool in the config file or called are  changed, the volumes list will be updated for this new disk_pool.

**Comment by trenxiaof** [comment]
@bjhuangr please help review

**Review by bjhuangr — APPROVED** [review]

**Pull request merged**